File: scripts/fetch-reviews.js
import { readFileSync, writeFileSync } from 'fs';
import fetch from 'node-fetch';

const API_KEY = process.env.GOOGLE_PLACES_API_KEY || '';
const businesses = JSON.parse(readFileSync('../data/businesses.json', 'utf-8'));

async function fetchReviews(placeId) {
  const url = `https://maps.googleapis.com/maps/api/place/details/json?place_id=${placeId}&fields=reviews&key=${API_KEY}`;
  try {
    const res = await fetch(url);
    const data = await res.json();
    if (data.status !== 'OK' || !data.result?.reviews) {
      console.error(`  ${data.status || 'No reviews'} for ${placeId}`);
      return [];
    }
    return data.result.reviews.slice(0, 5).map(r => ({
      author: r.author_name || 'Anonymous',
      rating: r.rating || 5,
      text: r.text || '',
      time: r.relative_time_description || '',
      profilePhoto: r.profile_photo_url || ''
    }));
  } catch (err) {
    console.error(`  Error for ${placeId}:`, err.message);
    return [];
  }
}

async function main() {
  const reviews = {};
  let fetched = 0;

  for (const b of businesses) {
    if (!b.place_id) continue;
    console.log(`Fetching reviews for: ${b.name}`);
    const r = await fetchReviews(b.place_id);
    if (r.length) {
      reviews[b.slug] = r;
      fetched++;
    }
    await new Promise(resolve => setTimeout(resolve, 200));
  }

  writeFileSync('../data/reviews.json', JSON.stringify(reviews, null, 2));
  console.log(`\nDone. Fetched reviews for ${fetched}/${businesses.length} businesses.`);
}

main();

File: scripts/fetch_reviews.py
"""
Fetch Google reviews for all businesses and save to data/reviews.json.

Usage:
    python scripts/fetch_reviews.py YOUR_GOOGLE_API_KEY

Requires: requests (pip install requests)
Uses the Google Places API (New) to fetch up to 5 reviews per business.
"""

import json
import sys
import os
import time

try:
    import requests
except ImportError:
    print("Error: 'requests' package required. Install with: pip install requests")
    sys.exit(1)


def fetch_reviews(place_id, api_key):
    """Fetch reviews for a single place using Google Places API (New)."""
    url = f"https://places.googleapis.com/v1/places/{place_id}"
    headers = {
        "X-Goog-Api-Key": api_key,
        "X-Goog-FieldMask": "reviews"
    }
    resp = requests.get(url, headers=headers, timeout=15)
    if resp.status_code != 200:
        print(f"  Warning: API returned {resp.status_code} for {place_id}")
        return []

    data = resp.json()
    reviews = []
    for r in data.get("reviews", []):
        reviews.append({
            "author": r.get("authorAttribution", {}).get("displayName", "Anonymous"),
            "profile_photo": r.get("authorAttribution", {}).get("photoUri", ""),
            "rating": r.get("rating", 5),
            "text": r.get("text", {}).get("text", ""),
            "time": r.get("relativePublishTimeDescription", "")
        })
    return reviews


def main():
    if len(sys.argv) < 2:
        print("Usage: python scripts/fetch_reviews.py YOUR_GOOGLE_API_KEY")
        sys.exit(1)

    api_key = sys.argv[1]
    script_dir = os.path.dirname(os.path.abspath(__file__))
    project_dir = os.path.dirname(script_dir)
    businesses_path = os.path.join(project_dir, "data", "businesses.json")
    reviews_path = os.path.join(project_dir, "data", "reviews.json")

    with open(businesses_path, "r", encoding="utf-8") as f:
        businesses = json.load(f)

    # Load existing reviews to preserve them if fetch fails
    existing = {}
    if os.path.exists(reviews_path):
        with open(reviews_path, "r", encoding="utf-8") as f:
            existing = json.load(f)

    all_reviews = dict(existing)
    total = len(businesses)

    for i, biz in enumerate(businesses):
        slug = biz["slug"]
        place_id = biz.get("place_id", "")
        print(f"[{i+1}/{total}] {biz['name']}...", end=" ")

        if not place_id:
            print("no place_id, skipping")
            continue

        reviews = fetch_reviews(place_id, api_key)
        if reviews:
            all_reviews[slug] = reviews
            print(f"{len(reviews)} reviews")
        else:
            print("no reviews")

        # Rate limiting - stay under quota
        time.sleep(0.2)

    with open(reviews_path, "w", encoding="utf-8") as f:
        json.dump(all_reviews, f, indent=2, ensure_ascii=False)

    print(f"\nDone! Saved reviews for {len(all_reviews)} businesses to data/reviews.json")


if __name__ == "__main__":
    main()
